### PR TITLE
Support .NET SDK Container Build

### LINKED ISSUE #3808: Enable debugging support for .NET native container builds
Starting with .NET 7, the .NET SDK enables [containerizing .NET applications without a Dockerfile](https://devblogs.microsoft.com/dotnet/announcing-builtin-container-support-for-the-dotnet-sdk/) via a simple `dotnet publish` command. Subsequent releases of this capability have focused on integrating the package into more project types - the 7.0.200 release of the SDK in February will include the package for Web SDK projects (those that use ASP.NET Core), and we plan to extend this to all .NET project types in the spring release of 7.0.300.

It would be amazing if we could integrate this package into the tooling for this extension, letting developers continue the easy containerization experience all the way up to publishing their app to the cloud provider even inside their IDE.

Conceretely, it would be great if the onboarding experience for .NET debugging support didn't require the use of scaffolded Dockerfiles, and could generate `publish` commands instead of `docker build` commands for the prerequisite build tasks.

### PR DESCRIPTION
**Pull request by alexyaang**
This PR closes #3808

Tests
- [x] Debug a .NET project with .NET SDK Container
- [x] Debug a .NET project with Dockerfile (check for regression)
- [x] Debug a python project with Dockerfile (check for regression)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -636,6 +636,11 @@
             {
                 "type": "docker",
                 "label": "Docker: Debug in Container",
+                "languages": [
+                    "csharp",
+                    "razor",
+                    "aspnetcorerazor"
+                ],
```
Comment (alexyaang) [inline package.json:643]: should we add python & node.js here too? so when you're on a .py file or .js file and presses `f5`, you'll be able to see our debugger. I was thinking we can do that in a separate PR.

File: src/debugging/DockerDebugConfigurationProvider.ts [hunk 2]
```diff
@@ -42,21 +44,14 @@ export class DockerDebugConfigurationProvider implements DebugConfigurationProvi
...
-            .then((result) => {
-                if (result === add) {
-                    /* eslint-disable-next-line @typescript-eslint/no-floating-promises */
-                    commands.executeCommand('vscode-docker.configure');
-                }
-            });
 
-        return [];
+        return callWithTelemetryAndErrorHandling(
+            'docker-provideDebugConfigurations', //TODO: change this later
```
Comment (alexyaang) [inline src/debugging/DockerDebugConfigurationProvider.ts:49]: what should we name this? is `docker-provideDebugConfigurations` sufficient?

File: src/utils/StateUtils.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ext } from '../extensionVariables';
+
+export async function getFromWorkspaceState<T>(key: string): Promise<T> {
```
Comment (alexyaang) [inline src/utils/StateUtils.ts:8]: do we think having a util class for workplace momentos is needed? I thought it would be nice to have a util class that handles it, but we can go either way.

File: package.json (hunk 1 above)
Comment (bwateratmsft) [reply package.json:643]: I think another PR is fine

File: resources/netCore/GetProjectProperties.targets [hunk 4]
```diff
@@ -1,11 +1,16 @@
 <Project>
-    <!-- All the relevant info is in root-level PropertyGroups, so there are no dependent targets to make this work -->
-    <Target Name="GetProjectProperties">
+    <PropertyGroup>
+        <GetProjectPropertiesDependsOn Condition=" '$(SDKContainerSupportEnabled)' == 'true' ">$(GetProjectPropertiesDependsOn);ComputeContainerConfig;</GetProjectPropertiesDependsOn>
```
Comment (bwateratmsft) [inline resources/netCore/GetProjectProperties.targets:3]: This file still works when SDK container support is not enabled, right?

File: src/debugging/DockerDebugConfigurationProvider.ts (hunk 2 above)
Comment (bwateratmsft) [reply src/debugging/DockerDebugConfigurationProvider.ts:49]: The telemetry name will be prepended with `vscode-docker.` automatically so I think we can at least remove the `docker-`.

File: src/debugging/DockerDebugConfigurationProvider.ts [hunk 5]
```diff
@@ -75,10 +70,15 @@ export class DockerDebugConfigurationProvider implements DebugConfigurationProvi
                     }
                 }
 
-                if (debugConfiguration.type === undefined) {
-                    // If type is undefined, they may be doing F5 without creating any real launch.json, which won't work
-                    // VSCode subsequently will call provideDebugConfigurations which will show an error message
-                    return null;
+                if (debugConfiguration.type === undefined &&
```
Comment (bwateratmsft) [inline src/debugging/DockerDebugConfigurationProvider.ts:73]: Should we look specifically for `undefined` or should any falsy value trigger this?

File: src/debugging/DockerDebugConfigurationProvider.ts [hunk 6]
```diff
@@ -170,4 +170,51 @@ export class DockerDebugConfigurationProvider implements DebugConfigurationProvi
             }
         }
     }
+
+    /**
+     * Copies the first debug configuration from the array to the second one.
+     */
+    private copyDebugConfiguration(fromArray: DockerDebugConfiguration[], to: DockerDebugConfiguration): void {
```
Comment (bwateratmsft) [inline src/debugging/DockerDebugConfigurationProvider.ts:177]: Does it not work to just do something like `to = fromArray[0]`?

Comment (bwateratmsft) [reply src/debugging/DockerDebugConfigurationProvider.ts:177]: (Or, the equivalent in the above function)

File: src/debugging/DockerDebugConfigurationProvider.ts [hunk 7]
```diff
@@ -170,4 +170,51 @@ export class DockerDebugConfigurationProvider implements DebugConfigurationProvi
...
+        if (csProjUris) {
+            return await netSdkDebugHelper.provideDebugConfigurations(
+                {
+                    actionContext,
+                    dockerfile: undefined,
+                    folder: folder
+                },
+                {
+                    appProject: csProjUris[0]?.absoluteFilePath || '',
+                });
```
Comment (bwateratmsft) [inline src/debugging/DockerDebugConfigurationProvider.ts:211]: Very minor nit

```suggestion
                }
            );
```

File: src/debugging/DockerDebugConfigurationProvider.ts [hunk 8]
```diff
@@ -170,4 +170,51 @@ export class DockerDebugConfigurationProvider implements DebugConfigurationProvi
...
+        }
+        else {
```
Comment (bwateratmsft) [inline src/debugging/DockerDebugConfigurationProvider.ts:213]: Very minor nit

```suggestion
        } else {
```

File: src/debugging/DockerDebugConfigurationProvider.ts [hunk 9]
```diff
@@ -170,4 +170,51 @@ export class DockerDebugConfigurationProvider implements DebugConfigurationProvi
...
+            // for now, we scaffold docker files
+            await commands.executeCommand('vscode-docker.configure');
+            return [];
```
Comment (bwateratmsft) [inline src/debugging/DockerDebugConfigurationProvider.ts:216]: Should we throw a `UserCancelledError` instead?

File: src/debugging/netSdk/NetSdkDebugHelper.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import { commands } from "vscode";
+import { NetChooseBuildTypeContext, netContainerBuild } from "../../scaffolding/wizard/net/NetContainerBuild";
+import { unresolveWorkspaceFolder } from "../../utils/resolveVariables";
+import { DockerDebugScaffoldContext } from "../DebugHelper";
+import { DockerDebugConfiguration } from "../DockerDebugConfigurationProvider";
+import { NetCoreDebugHelper, NetCoreDebugScaffoldingOptions } from "../netcore/NetCoreDebugHelper";
+
+const netContainerSdkBuildSymbol = 'dotnet-container-sdk: debug';
```
Comment (bwateratmsft) [inline src/debugging/netSdk/NetSdkDebugHelper.ts:8]: Nit: File-level constants usually get PascalCase
```suggestion
const NetContainerSdkBuildSymbol = 'dotnet-container-sdk: debug';
```

File: src/debugging/netSdk/NetSdkDebugHelper.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    /**
+     * Checks if the launch task is using the .NET SDK Container build
+     * @param preLaunchTask
+     * @returns true if the launch task is using the .NET SDK Container build
+     *          false otherwise
+     */
+    public isDotnetSdkBuild(preLaunchTask: string): boolean {
+        return preLaunchTask && preLaunchTask === netContainerSdkBuildSymbol;
```
Comment (bwateratmsft) [inline src/debugging/netSdk/NetSdkDebugHelper.ts:47]: This is valid but you could also do this:

```suggestion
        return preLaunchTask === netContainerSdkBuildSymbol;
```

File: src/debugging/netSdk/NetSdkDebugHelper.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+        const netCoreBuildContext: NetChooseBuildTypeContext = {
+            ...context.actionContext,
+            scaffoldType: 'debugging',
+            workspaceFolder: context.folder,
+        };
+
+        await netContainerBuild(netCoreBuildContext);
+
+        if (netCoreBuildContext?.containerBuildOptions === 'Use .NET SDK') {
```
Comment (bwateratmsft) [inline src/debugging/netSdk/NetSdkDebugHelper.ts:23]: Use the constant declaration in NetSdkChooseBuildStep.ts rather than cloning a new string

File: src/debugging/netSdk/NetSdkDebugHelper.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import { DockerDebugScaffoldContext } from "../DebugHelper";
+import { DockerDebugConfiguration } from "../DockerDebugConfigurationProvider";
+import { NetCoreDebugHelper, NetCoreDebugScaffoldingOptions } from "../netcore/NetCoreDebugHelper";
+
+const netContainerSdkBuildSymbol = 'dotnet-container-sdk: debug';
+export class NetSdkDebugHelper extends NetCoreDebugHelper {
+
+    public async provideDebugConfigurations(context: DockerDebugScaffoldContext, options?: NetCoreDebugScaffoldingOptions): Promise<DockerDebugConfiguration[]> {
+
+        const providers: DockerDebugConfiguration[] = [];
```
Comment (bwateratmsft) [inline src/debugging/netSdk/NetSdkDebugHelper.ts:13]:
```suggestion
        const configurations: DockerDebugConfiguration[] = [];
```

File: src/scaffolding/wizard/net/NetSdkChooseBuildStep.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ *--------------------------------------------------------------------------------------------*/
+
+import { IAzureQuickPickItem } from '@microsoft/vscode-azext-utils';
+import * as vscode from 'vscode';
+import { getFromWorkspaceState, updateWorkspaceState } from '../../../utils/StateUtils';
+import { TelemetryPromptStep } from '../TelemetryPromptStep';
+import { NetChooseBuildTypeContext } from './NetContainerBuild';
+
+export const NetContainerBuildOptions = [
+    'Use a Dockerfile',
```
Comment (bwateratmsft) [inline src/scaffolding/wizard/net/NetSdkChooseBuildStep.ts:13]: These should be localized

File: src/scaffolding/wizard/net/NetContainerBuild.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    ];
+
+    const wizard = new AzureWizard<NetChooseBuildTypeContext>(wizardContext as NetChooseBuildTypeContext, {
+        promptSteps: promptSteps,
+        title: vscode.l10n.t('Initialize for Debugging'),
+    });
+
+    await wizard.prompt();
+    await wizard.execute();
+
```
Comment (bwateratmsft) [inline src/scaffolding/wizard/net/NetContainerBuild.ts:31]: Nit: extra newline
```suggestion
```

File: src/tasks/netSdk/NetSdkRunTaskProvider.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+import { DockerTaskProvider } from '../DockerTaskProvider';
+import { DockerTaskExecutionContext } from '../TaskHelper';
+import { NET_SDK_RUN_TASK_SYMBOL, NetSdkTaskHelper } from './NetSdkTaskHelper';
+
+const netSdkDebugTaskName = 'debug';
+export class NetSdkRunTaskProvider extends DockerTaskProvider {
+
+    public constructor(protected readonly helper: NetSdkTaskHelper) { super(NET_SDK_RUN_TASK_SYMBOL, undefined); }
+
+    provideTasks(token: CancellationToken): ProviderResult<Task[]> {
```
Comment (bwateratmsft) [inline src/tasks/netSdk/NetSdkRunTaskProvider.ts:17]: Nit for clarity
```suggestion
    public provideTasks(token: CancellationToken): Task[] {
```

File: src/tasks/netSdk/NetSdkTaskHelper.ts [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    | 'x64'
+    | 'x86'
+    | 'arm64'
+    | 'arm'
+    | 'ppc64le'
+    | 'mips64'
+    | 's390x'
+    | string;
+
+export const NET_SDK_RUN_TASK_SYMBOL = 'dotnet-container-sdk';
```
Comment (bwateratmsft) [inline src/tasks/netSdk/NetSdkTaskHelper.ts:35]: Would prefer PascalCase for these constants

File: src/tasks/netSdk/NetSdkRunTaskProvider.ts [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+        // we need to initialize a task first so we can pass it into `DockerPseudoterminal`
+        const task = new Task(
```
Comment (bwateratmsft) [inline src/tasks/netSdk/NetSdkRunTaskProvider.ts:20]: This feels weird. Is there no way to use the same task?

File: src/tasks/netSdk/NetSdkTaskHelper.ts [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+const IMAGE_TAG = 'dev'; // intentionally default to dev tag for phase 1 of this feature
+const ERR_MSG_NO_WORKSPACE_FOLDER = l10n.t(`Unable to determine task scope to execute task ${NET_SDK_RUN_TASK_SYMBOL}. Please open a workspace folder.`);
+
+export class NetSdkTaskHelper {
```
Comment (bwateratmsft) [inline src/tasks/netSdk/NetSdkTaskHelper.ts:39]: Since this class doesn't implement `TaskHelper` it should be named something else

File: src/tasks/netSdk/NetSdkTaskHelper.ts [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+        const quotedArgs = Shell.getShellOrDefault().quote(command.args);
+        const commandLine = [client.commandName, ...quotedArgs].join(' ');
+
+        return commandLine;
+    }
+
+    public async inferProjPath(context: IActionContext, folder: WorkspaceFolder): Promise<string> {
+        const noProjectFileErrMessage = l10n.t('No .csproj file could be found.');
+        const item = await quickPickProjectFileItem(context, undefined, folder, noProjectFileErrMessage);
+        return item.absoluteFilePath || '';
```
Comment (bwateratmsft) [inline src/tasks/netSdk/NetSdkTaskHelper.ts:91]: Why the `|| ''`? Is that to convince TypeScript that there's definitely a value being returned?

File: src/utils/StateUtils.ts (hunk 3 above)
Comment (bwateratmsft) [reply src/utils/StateUtils.ts:8]: Maybe overkill since it's a one-liner that replaces a one-liner, but I don't feel strongly one way or the other.

File: resources/netCore/GetProjectProperties.targets (hunk 4 above)
Comment (alexyaang) [reply resources/netCore/GetProjectProperties.targets:3]: yeah,`$(OutputPath)$(AssemblyName).dll` will remain unchanged

File: src/debugging/DockerDebugConfigurationProvider.ts (hunk 5 above)
Comment (alexyaang) [reply src/debugging/DockerDebugConfigurationProvider.ts:73]: that's a good question. we can do `!debugConfiguration.type`  and make it so if `type` is an empty string and it still triggers it? it just depends on how flexible we want the debug experience to be.

File: src/tasks/netSdk/NetSdkTaskHelper.ts (hunk 20 above)
Comment (alexyaang) [reply src/tasks/netSdk/NetSdkTaskHelper.ts:91]: my intention was that if `item.absoluteFilePath` is `undefined` then at least it'll return a empty string rather than throw an error? but after looking at `quickPickProjectFileItem` closer, it handles the case where a `fileItem` is not found so I'll remove the `|| ''`. good catch!

File: src/tasks/netSdk/NetSdkRunTaskProvider.ts (hunk 18 above)
Comment (alexyaang) [reply src/tasks/netSdk/NetSdkRunTaskProvider.ts:20]: I tried a few ways to rewrite it but I wasn't able to...

File: src/tasks/netSdk/NetSdkTaskHelper.ts (hunk 17 above)
Comment (alexyaang) [reply src/tasks/netSdk/NetSdkTaskHelper.ts:35]: sounds good!

File: src/tasks/netSdk/NetSdkTaskHelper.ts (hunk 19 above)
Comment (bwateratmsft) [reply src/tasks/netSdk/NetSdkTaskHelper.ts:39]: Or `NetSdkTaskUtils`? They could also be pure functions instead of class members

File: src/debugging/netSdk/NetSdkDebugHelper.ts (hunk 11 above)
Comment (alexyaang) [reply src/debugging/netSdk/NetSdkDebugHelper.ts:47]: looks good !

File: src/debugging/DockerDebugConfigurationProvider.ts (hunk 6 above)
Comment (alexyaang) [reply src/debugging/DockerDebugConfigurationProvider.ts:177]: good suggestion, it does work actually

File: src/tasks/netSdk/NetSdkRunTaskProvider.ts (hunk 18 above)
Comment (alexyaang) [reply src/tasks/netSdk/NetSdkRunTaskProvider.ts:20]: update: was able to do a little magic to avoid this w/ @bwateratmsft 's help

File: src/tasks/netSdk/NetSdkTaskHelper.ts (hunk 19 above)
Comment (alexyaang) [reply src/tasks/netSdk/NetSdkTaskHelper.ts:39]: @danegsta any preference on this?

File: src/tasks/netSdk/NetSdkTaskHelper.ts [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+            default:
+                return architecture;
+        }
+    }
+
+    /**
+     * This methods returns the mount for the remote debugger ONLY as the SDK built container will have
+     * everything it needs to run the app already inside.
+     */
+    private async getRemoteDebuggerMount(): Promise<RunContainerBindMount[] | undefined> {
```
Comment (danegsta) [inline src/tasks/netSdk/NetSdkTaskHelper.ts:122]: Consider adding a shared utility function that can be re-used both here and in NetCoreTaskHelper.ts which takes an OSType and returns the appropriate debugger, since the logic for setting a debugger path should be the same between the two scenarios.

File: src/tasks/netSdk/NetSdkTaskHelper.ts (hunk 19 above)
Comment (danegsta) [reply src/tasks/netSdk/NetSdkTaskHelper.ts:39]: I like naming the file either `netSdkTaskUtils` or `netSdkUtils` and agree with Brandon that it makes sense to export the individual functions instead of a class.

File: src/tasks/netSdk/NetSdkTaskHelper.ts [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+        const volumes: DockerContainerVolume[] = [];
+        const isLinux = await getDockerOSType() === 'linux';
```
Comment (danegsta) [inline src/tasks/netSdk/NetSdkTaskHelper.ts:124]: I'd suggest using isWindows as the primary condition you check against (similar to the behavior in NetCoreTaskHelper.ts), since there could theoretically be additional Docker OS types, such as darwin, but they will generally match Linux conventions. Treating Windows as the exceptional case is usually safer.

File: src/tasks/netSdk/NetSdkTaskHelper.ts [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+        const command = await client.runContainer({
+            detached: true,
+            publishAllPorts: true,
+            name: getDefaultContainerName(folderName.name, IMAGE_TAG),
+            environmentVariables: {},
+            removeOnExit: true,
+            imageRef: getDefaultImageName(folderName.name, IMAGE_TAG),
+            labels: defaultVsCodeLabels,
+            mounts: await this.getRemoteDebuggerMount(),
+            customOptions: '--expose 8080',
```
Comment (danegsta) [inline src/tasks/netSdk/NetSdkTaskHelper.ts:78]: We should add a new `expose` property to `RunContainerCommandOptions` rather than falling back to customOptions. Ideally you could specify `expose: [8080],` here.

File: src/debugging/netSdk/NetSdkDebugHelper.ts [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+        return configurations;
+    }
+
+    /**
+     * Checks if the launch task is using the .NET SDK Container build
+     * @param preLaunchTask
+     * @returns true if the launch task is using the .NET SDK Container build
+     *          false otherwise
+     */
+    public isDotnetSdkBuild(preLaunchTask: string): boolean {
```
Comment (danegsta) [inline src/debugging/netSdk/NetSdkDebugHelper.ts:47]:
```suggestion
    public isDotNetSdkBuild(preLaunchTask: string): boolean {
```

File: src/tasks/netSdk/NetSdkTaskHelper.ts (hunk 19 above)
Comment (alexyaang) [reply src/tasks/netSdk/NetSdkTaskHelper.ts:39]: will do!

File: src/tasks/netSdk/NetSdkTaskHelper.ts (hunk 21 above)
Comment (alexyaang) [reply src/tasks/netSdk/NetSdkTaskHelper.ts:122]: I like the idea but after playing around with it, I think this might make the code in `NetCoreTaskHelper.ts` more confusing. We could add a method to `TaskHelper` called  `addDebuggerVolumeWithoutConflict()` and it would make `inferVolumes()` look like this. Which IMO is more confusing than what we have now. 

``` TypeScript
...
addVolumeWithoutConflicts(volumes, appVolume);
addVolumeWithoutConflicts(volumes, srcVolume);
addDebuggerVolumeWithoutConflicts(volumes, osType);
if (await fse.pathExists(nugetRootVolume.localPath)) {
    addVolumeWithoutConflicts(volumes, nugetRootVolume);
}
...
```

File: src/debugging/DockerDebugConfigurationProvider.ts (hunk 5 above)
Comment (alexyaang) [reply src/debugging/DockerDebugConfigurationProvider.ts:73]: @danegsta any thoughts on this btw?

File: src/tasks/netSdk/netSdkTaskUtils.ts [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+    const volumes: DockerContainerVolume[] = [];
+    const isWindows = await getDockerOSType() === 'windows';
+
+    const debuggerVolume: DockerContainerVolume = {
+        localPath: vsDbgInstallBasePath,
+        containerPath: isWindows ? 'C:\\remote_debugger' : '/remote_debugger',
+        permissions: 'ro'
+    };
+
+    addVolumeWithoutConflicts(volumes, debuggerVolume);
```
Comment (danegsta) [inline src/tasks/netSdk/netSdkTaskUtils.ts:129]: Why not return a RunContainerBindMount directly? You're always starting with an empty set of volumes here, so addVolumeWithoutConflicts/getMounts aren't really doing anything here.

Comment (danegsta) [reply src/tasks/netSdk/netSdkTaskUtils.ts:129]: Honestly, I don't love that we go back and forth between DockerContainerVolume and RunContainerBindMount here and in inferVolumes; I think I'd like it more if we normalized the volumes to RunContainerBindMount[] early and did all our volume logic with that type instead.

File: src/runtimes/docker/contracts/ContainerClient.ts [hunk 26]
```diff
@@ -744,6 +744,10 @@ export type RunContainerCommandOptions = CommonCommandOptions & {
      * Optional command to use in starting the container
      */
     command?: Array<string> | string;
+    /**
+     * Optional expose ports for the container
+     */
+    exposes?: Array<number>;
```
Comment (bwateratmsft) [inline src/runtimes/docker/contracts/ContainerClient.ts:750]: I don't feel strongly one way or the other, but maybe "expose"? Most of the other array arguments have a plural, but "exposes" isn't a plural...I don't know that there is a plural of "expose".

Comment (danegsta) [reply src/runtimes/docker/contracts/ContainerClient.ts:750]: Maybe exposePorts?

Comment (alexyaang) [reply src/runtimes/docker/contracts/ContainerClient.ts:750]: good point! and I think`exposePorts` make the most sense to me

File: src/runtimes/docker/clients/DockerClientBase/withDockerExposesArg.ts [hunk 27]
```diff
@@ -0,0 +1,10 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { withNamedArg } from "../../utils/commandLineBuilder";
+
+export function withDockerExposesArg(ports?: Array<number>) {
+    return withNamedArg('--expose', ports.toString(), { shouldQuote: false });
```
Comment (danegsta) [inline src/runtimes/docker/clients/DockerClientBase/withDockerExposesArg.ts:9]: You'll want to keep the ports argument as an array, otherwise you'll get an error if multiple ports are passed in:

```suggestion
    return withNamedArg('--expose', (ports || []).map(port => port.toString()), { shouldQuote: false });
```

Comment (alexyaang) [reply src/runtimes/docker/clients/DockerClientBase/withDockerExposesArg.ts:9]: that's right... rookie mistake :disappointed_relieved:

File: src/debugging/DockerDebugConfigurationProvider.ts (hunk 5 above)
Comment (danegsta) [reply src/debugging/DockerDebugConfigurationProvider.ts:73]: What's the actual value of debugConfiguration in the event there isn't an existing launch.json? Is it just `{}`?

File: src/runtimes/docker/clients/DockerClientBase/withDockerExposesArg.ts (hunk 27 above)
Comment (danegsta) [reply src/runtimes/docker/clients/DockerClientBase/withDockerExposesArg.ts:9]: It's an easy mistake to make with any kind of array data; happens to us all.

File: src/debugging/DockerDebugConfigurationProvider.ts (hunk 5 above)
Comment (alexyaang) [reply src/debugging/DockerDebugConfigurationProvider.ts:73]: just double checked and yeah exactly, debugConfiguration would be `{}`

**Review by bwateratmsft — DISMISSED** [review]
Looks good to me. @danegsta can you also take a final look?

File: src/tasks/netSdk/netSdkTaskUtils.ts [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+    const noProjectFileErrMessage = l10n.t('No .csproj file could be found.');
+    const item = await quickPickProjectFileItem(context, undefined, folder, noProjectFileErrMessage);
+    return item.absoluteFilePath;
+}
+
+/**
+ * This method normalizes the Docker OS type to match the .NET Core SDK conventions.
+ * {@link https://learn.microsoft.com/en-us/dotnet/core/rid-catalog}
+ */
+async function normalizeOsToRid(): Promise<'linux' | 'win'> {
```
Comment (danegsta) [inline src/tasks/netSdk/netSdkTaskUtils.ts:94]:
```suggestion
async function normalizeOsToRidOs(): Promise<'linux' | 'win'> {
```

File: src/tasks/netSdk/netSdkTaskUtils.ts [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+    const dockerOsType = await getDockerOSType();
+    return dockerOsType === 'windows' ? 'win' : 'linux';
+}
+
+/**
+ * This method normalizes the native architecture to match the .NET Core SDK conventions.
+ * {@link https://learn.microsoft.com/en-us/dotnet/core/rid-catalog}
+ */
+async function normalizeArchitectureToRid(): Promise<RidCpuArchitecture> {
```
Comment (danegsta) [inline src/tasks/netSdk/netSdkTaskUtils.ts:103]:
```suggestion
async function normalizeArchitectureToRidArchitecture(): Promise<RidCpuArchitecture> {
```

File: src/tasks/netSdk/netSdkTaskUtils.ts (hunk 28 above)
Comment (alexyaang) [reply src/tasks/netSdk/netSdkTaskUtils.ts:94]: good suggestions

File: src/tasks/netSdk/netSdkTaskUtils.ts (hunk 29 above)
Comment (alexyaang) [reply src/tasks/netSdk/netSdkTaskUtils.ts:103]: good suggestion as well

File: src/tasks/netSdk/netSdkTaskUtils.ts [hunk 30]
```diff
@@ -0,0 +1,128 @@
...
+    return item.absoluteFilePath;
+}
+
+/**
+ * This method normalizes the Docker OS type to match the .NET Core SDK conventions.
+ * {@link https://learn.microsoft.com/en-us/dotnet/core/rid-catalog}
+ */
+async function normalizeOsToRid(): Promise<'linux' | 'win'> {
+    const dockerOsType = await getDockerOSType();
+    return dockerOsType === 'windows' ? 'win' : 'linux';
```
[the remark this answers is not in the document]
Comment (alexyaang) [reply src/tasks/netSdk/netSdkTaskUtils.ts:96]: That's a good point. Let's discuss and make a follow up PR to address the issue.

**Review by bwateratmsft — APPROVED** [review]

File: src/tasks/netSdk/netSdkTaskUtils.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (baronfel) [reply src/tasks/netSdk/netSdkTaskUtils.ts:96]: We don't currently support inferring musl-based images in the SDK container tasks but would like to for 7.0.400 or 8.0.100 - if we are able to surface that in the Targets/props would you be able to consume it easily?
